File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool isPossible(vector<int>& v, int m ,int k, int mid){
        if(m==0) return true;
        int count=0;
        for(int i=0;i<v.size();i++){
            if(v[i]-mid<=0){
                count++;
                  if(count==k){
                    m--;
                    count=0;
                }
            }
            else count=0;
            if(m<=0) return true;
        }
        return false;
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        int n = bloomDay.size();
        int max_ele = INT_MIN;
        for(int i: bloomDay){
            if(i>max_ele) max_ele=i;
        }
        int start=0, end= max_ele,mid,ans=-1;
        while(start<=end){
            mid=start+(end-start)/2;
            if(isPossible(bloomDay,m,k,mid)){
                ans=mid;
                end=mid-1;
            }
            else start=mid+1;
        }
        return ans;
    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        int minDay = 0;
        int maxDay=0;
        int ans=-1;
        int n = bloomDay.length;
        for(int day: bloomDay){
            minDay=Math.min(minDay,day);
            maxDay=Math.max(maxDay,day);
        }
        while(minDay<=maxDay){
            int mid =(minDay+maxDay)/2;
            if(solve(bloomDay,m,k,mid,n)){
                ans=mid;
                maxDay=mid-1;
            }
            else minDay=mid+1;
        }
        return ans;
    }
    private boolean solve(int[] bloomDay, int m, int k,int mid,int n){
        int count=0;
        int flowers=0;
        for(int i=0;i<n;i++){
            if(bloomDay[i]<=mid){
                flowers++;
                if(flowers==k){
                    count++;
                    flowers=0;
                }
            }
            else{
                flowers=0;
            }
        }
        return count>=m;
    }
}